# Spotify Track Analysis
## Section 1 — Data Loading and Initial Inspection

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

Matplotlib is building the font cache; this may take a moment.


In [2]:
# Load raw dataset
df = pd.read_csv("../data/raw/spotify_tracks.csv")

# Sample down to 50,000 rows
df = df.sample(n=50000, random_state=42)

# Reset index after sampling
df = df.reset_index(drop=True)

print("Dataset loaded successfully!")
print(f"Shape: {df.shape}")

Dataset loaded successfully!
Shape: (50000, 21)


In [3]:
# First look at the data
df.head()

,Unnamed: 0,track_id,artists,album_name,track_name,popularity,duration_ms,explicit,danceability,energy,...,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature,track_genre
0,113186,6KwkVtXm8OUp2XffN5k7lY,Hillsong Worship,No Other Name,No Other Name,50,440247,False,0.369,0.598,...,-6.984,1,0.0304,0.00511,0.000000,0.176,0.0466,148.014,4,world-music
1,42819,2dp5I5MJ8bQQHDoFaNRFtX,Internal Rot,Grieving Birth,Failed Organum,11,93933,False,0.171,0.997,...,-3.586,1,0.1180,0.00521,0.801000,0.420,0.0294,122.223,4,grindcore
2,59311,5avw06usmFkFrPjX8NxC40,Zhoobin Askarieh;Ali Sasha,Noise A Noise 20.4-1,"Save the Trees, Pt. 1",0,213578,False,0.173,0.803,...,-10.071,0,0.1440,0.61300,0.001910,0.195,0.0887,75.564,3,iranian
3,91368,75hT0hvlESnDJstem0JgyR,Bryan Adams,All I Want For Christmas Is You,Merry Christmas,0,151387,False,0.683,0.511,...,-5.598,1,0.0279,0.40600,0.000197,0.111,0.5980,109.991,3,rock
4,61000,4bY2oZGA5Br3pTE1Jd1IfY,Nogizaka46,バレッタ TypeD,月の大きさ,57,236293,False,0.555,0.941,...,-3.294,0,0.0481,0.48400,0.000000,0.266,0.8130,92.487,4,j-idol


In [4]:
# Column names and data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        50000 non-null  int64  
 1   track_id          50000 non-null  str    
 2   artists           49999 non-null  str    
 3   album_name        49999 non-null  str    
 4   track_name        49999 non-null  str    
 5   popularity        50000 non-null  int64  
 6   duration_ms       50000 non-null  int64  
 7   explicit          50000 non-null  bool   
 8   danceability      50000 non-null  float64
 9   energy            50000 non-null  float64
 10  key               50000 non-null  int64  
 11  loudness          50000 non-null  float64
 12  mode              50000 non-null  int64  
 13  speechiness       50000 non-null  float64
 14  acousticness      50000 non-null  float64
 15  instrumentalness  50000 non-null  float64
 16  liveness          50000 non-null  float64
 17  vale

In [5]:
# Basic descriptive statistics
df.describe()

,Unnamed: 0,popularity,duration_ms,danceability,energy,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,time_signature
count,50000.000000,50000.000000,5.000000e+04,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.00000,50000.000000,50000.000000,50000.0000
mean,56893.732100,33.256920,2.282052e+05,0.565963,0.641456,5.315200,-8.246604,0.636280,0.084398,0.313464,0.155164,0.21352,0.473878,122.211438,3.9055
std,32809.686522,22.283949,1.108710e+05,0.174015,0.251491,3.555894,5.011987,0.481074,0.105666,0.332277,0.308354,0.19040,0.259784,30.015696,0.4312
min,1.000000,0.000000,0.000000e+00,0.000000,0.000000,0.000000,-49.531000,0.000000,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.0000
25%,28761.750000,17.000000,1.736992e+05,0.456000,0.472000,2.000000,-10.034000,0.000000,0.035800,0.016500,0.000000,0.09820,0.256750,99.290000,4.0000
50%,56784.000000,35.000000,2.128000e+05,0.579000,0.685000,5.000000,-6.990000,1.000000,0.048700,0.167000,0.000046,0.13200,0.464000,122.037000,4.0000
75%,85149.000000,50.000000,2.616138e+05,0.694000,0.853000,8.000000,-5.002750,1.000000,0.084400,0.594000,0.049000,0.27200,0.683000,140.109250,4.0000
max,113999.000000,98.000000,5.237295e+06,0.985000,1.000000,11.000000,1.795000,1.000000,0.965000,0.996000,1.000000,1.00000,0.992000,243.372000,5.0000


In [6]:
# Check missing values per column
missing = df.isnull().sum()
print("Missing values per column:")
print(missing[missing > 0])

Missing values per column:
artists       1
album_name    1
track_name    1
dtype: int64


In [7]:
# Create a small Series by hand (required by rubric)
audio_features = pd.Series(
    [df['danceability'].mean(),
     df['energy'].mean(),
     df['tempo'].mean(),
     df['loudness'].mean(),
     df['valence'].mean()],
    index=['danceability', 'energy', 'tempo', 'loudness', 'valence'],
    name='average_values'
)
print("Audio Feature Averages:")
print(audio_features)

Audio Feature Averages:
danceability      0.565963
energy            0.641456
tempo           122.211438
loudness         -8.246604
valence           0.473878
Name: average_values, dtype: float64


In [8]:
# Create a small DataFrame by hand (required by rubric)
summary_df = pd.DataFrame({
    'feature': ['danceability', 'energy', 'tempo', 'loudness', 'valence'],
    'mean':    [df['danceability'].mean(),
                df['energy'].mean(),
                df['tempo'].mean(),
                df['loudness'].mean(),
                df['valence'].mean()],
    'median':  [df['danceability'].median(),
                df['energy'].median(),
                df['tempo'].median(),
                df['loudness'].median(),
                df['valence'].median()],
    'std':     [df['danceability'].std(),
                df['energy'].std(),
                df['tempo'].std(),
                df['loudness'].std(),
                df['valence'].std()]
})
print("Summary DataFrame:")
print(summary_df)

Summary DataFrame:
        feature        mean   median        std
0  danceability    0.565963    0.579   0.174015
1        energy    0.641456    0.685   0.251491
2         tempo  122.211438  122.037  30.015696
3      loudness   -8.246604   -6.990   5.011987
4       valence    0.473878    0.464   0.259784


In [9]:
# Check unique values in categorical columns
print("Unique genres:", df['track_genre'].nunique())
print("Explicit values:", df['explicit'].unique())

Unique genres: 114
Explicit values: [False  True]


In [10]:
# Check value counts for genre
print("Top 10 genres:")
print(df['track_genre'].value_counts().head(10))

Top 10 genres:
track_genre
jazz          474
emo           474
house         472
goth          472
electronic    467
party         466
dubstep       463
blues         462
rockabilly    462
idm           462
Name: count, dtype: int64


## Key Findings — Section 1

- The dataset contains **50,000 tracks** after sampling across **21 columns**.
- Key numeric features: danceability, energy, tempo, loudness, valence, popularity.
- Key categorical features: track_genre, explicit.
- Missing values found in: **artists, album_name, track_name (1 missing each)** — very clean dataset.
- Average danceability is **0.57** — most tracks are moderately danceable.
- Average energy is **0.64** — most tracks lean towards high energy.
- Average popularity is **33/100** — wide spread across tracks.
- There are **114 unique genres** in the dataset.
- Explicit tracks: **True/False** — can be used for grouped analysis.
- Top genres include jazz, emo, house, goth, and electronic.

## Section 2 — Cleaning and Transformation

In [11]:
# Drop the 3 rows with missing values in artists, album_name, track_name
df = df.dropna(subset=['artists', 'album_name', 'track_name'])

print("Shape after dropping missing rows:", df.shape)

Shape after dropping missing rows: (49999, 21)


In [12]:
# Drop the unnecessary 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

print("Columns after drop:")
print(df.columns.tolist())

Columns after drop:
['track_id', 'artists', 'album_name', 'track_name', 'popularity', 'duration_ms', 'explicit', 'danceability', 'energy', 'key', 'loudness', 'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'time_signature', 'track_genre']


In [13]:
# Convert duration from milliseconds to seconds (new column)
df['duration_sec'] = pd.to_numeric(df['duration_ms'] / 1000, errors='coerce')

print("Duration in seconds (first 5):")
print(df['duration_sec'].head())

Duration in seconds (first 5):
0    440.247
1     93.933
2    213.578
3    151.387
4    236.293
Name: duration_sec, dtype: float64


In [14]:
# Convert explicit column to integer (True=1, False=0)
df['explicit'] = df['explicit'].astype(int)

print("Explicit column after conversion:")
print(df['explicit'].value_counts())

Explicit column after conversion:
explicit
0    45778
1     4221
Name: count, dtype: int64


In [15]:
# Convert track_genre to categorical dtype
df['track_genre'] = df['track_genre'].astype('category')

print("track_genre dtype:", df['track_genre'].dtype)

track_genre dtype: category


In [16]:
# Derive new column using np.where
# Categorize songs as 'high' or 'low' popularity
df['popularity_tier'] = np.where(df['popularity'] >= 50, 'high', 'low')

print("Popularity tier counts:")
print(df['popularity_tier'].value_counts())

Popularity tier counts:
popularity_tier
low     37085
high    12914
Name: count, dtype: int64


In [17]:
# Derive new column using .apply
# Categorize tempo into slow, medium, fast
def categorize_tempo(tempo):
    if tempo < 90:
        return 'slow'
    elif tempo < 130:
        return 'medium'
    else:
        return 'fast'

df['tempo_category'] = df['tempo'].apply(categorize_tempo)

print("Tempo category counts:")
print(df['tempo_category'].value_counts())

Tempo category counts:
tempo_category
medium    24481
fast      18318
slow       7200
Name: count, dtype: int64


In [18]:
# Fill any remaining nulls in numeric columns with median
numeric_cols = df.select_dtypes(include='number').columns
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())

print("Missing values after cleaning:")
print(df.isnull().sum().sum())

Missing values after cleaning:
0


In [19]:
# Final check on cleaned dataframe
print("Final shape:", df.shape)
print("Dtypes:")
print(df.dtypes)

Final shape: (49999, 23)
Dtypes:
track_id                 str
artists                  str
album_name               str
track_name               str
popularity             int64
duration_ms            int64
explicit               int64
danceability         float64
energy               float64
key                    int64
loudness             float64
mode                   int64
speechiness          float64
acousticness         float64
instrumentalness     float64
liveness             float64
valence              float64
tempo                float64
time_signature         int64
track_genre         category
duration_sec         float64
popularity_tier          str
tempo_category           str
dtype: object


In [20]:
# Export cleaned data to data/processed/
df.to_csv("../data/processed/spotify_cleaned.csv", index=False)

print("Cleaned dataset exported successfully!")

Cleaned dataset exported successfully!


## Key Findings — Section 2

- Dropped **3 rows** with missing values in artists, album_name, track_name.
- Dropped **Unnamed: 0** column — not needed for analysis.
- Converted **duration_ms** to **duration_sec** for easier interpretation.
- Converted **explicit** from bool to int (1 = explicit, 0 = not explicit).
- Converted **track_genre** to categorical dtype for memory efficiency.
- Created **popularity_tier** column using np.where — high (>=50) vs low (<50).
- Created **tempo_category** column using .apply — slow/medium/fast.
- Filled remaining numeric nulls with median values.
- Cleaned dataset exported to **data/processed/spotify_cleaned.csv**.

## Section 3 — Descriptive Statistics and EDA

In [15]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Reload cleaned dataset
df = pd.read_csv("../data/processed/spotify_cleaned.csv")
print("Loaded successfully!")
print(df.shape)

Loaded successfully!
(49999, 23)


In [16]:
# Reload cleaned dataset
df = pd.read_csv("../data/processed/spotify_cleaned.csv")
print("Loaded successfully!")
print(df.shape)

Loaded successfully!
(49999, 23)


In [17]:
# Descriptive stats on key numeric columns
numeric_cols = ['popularity', 'danceability', 'energy', 
                'loudness', 'tempo', 'valence', 'acousticness',
                'instrumentalness', 'speechiness']

print(df[numeric_cols].describe().round(3))

       popularity  danceability     energy   loudness      tempo    valence  \
count   49999.000     49999.000  49999.000  49999.000  49999.000  49999.000   
mean       33.258         0.566      0.641     -8.247    122.211      0.474   
std        22.284         0.174      0.251      5.012     30.016      0.260   
min         0.000         0.000      0.000    -49.531      0.000      0.000   
25%        17.000         0.456      0.472    -10.034     99.288      0.256   
50%        35.000         0.579      0.685     -6.990    122.037      0.464   
75%        50.000         0.694      0.853     -5.002    140.110      0.683   
max        98.000         0.985      1.000      1.795    243.372      0.992   

       acousticness  instrumentalness  speechiness  
count     49999.000         49999.000    49999.000  
mean          0.313             0.155        0.084  
std           0.332             0.308        0.106  
min           0.000             0.000        0.000  
25%           0.016    

In [18]:
# Explicit stats on individual columns
for col in ['popularity', 'danceability', 'energy']:
    print(f"=== {col} ===")
    print(f"Mean:   {df[col].mean():.3f}")
    print(f"Median: {df[col].median():.3f}")
    print(f"Min:    {df[col].min():.3f}")
    print(f"Max:    {df[col].max():.3f}")
    print(f"Std:    {df[col].std():.3f}")
    print()

=== popularity ===
Mean:   33.258
Median: 35.000
Min:    0.000
Max:    98.000
Std:    22.284

=== danceability ===
Mean:   0.566
Median: 0.579
Min:    0.000
Max:    0.985
Std:    0.174

=== energy ===
Mean:   0.641
Median: 0.685
Min:    0.000
Max:    1.000
Std:    0.251



In [19]:
# Grouped stats by genre — mean popularity
genre_popularity = df.groupby('track_genre')['popularity'].agg(['mean', 'count', 'std'])
genre_popularity = genre_popularity.sort_values('mean', ascending=False)

print("Top 10 genres by average popularity:")
print(genre_popularity.head(10).round(3))

Top 10 genres by average popularity:
                     mean  count     std
track_genre                             
pop-film           58.450    440  11.196
k-pop              56.704    446  16.423
chill              53.528    424  15.021
sad                52.360    405  11.649
grunge             49.403    444  19.408
indian             49.273    436  11.493
anime              48.693    423  12.337
sertanejo          47.899    435   4.334
emo                47.705    474  17.701
progressive-house  47.160    406  19.535


In [20]:
# Grouped stats by genre — multiple features
genre_features = df.groupby('track_genre')[['danceability', 'energy', 
                                             'valence', 'popularity']].agg(['mean', 'count'])
print("Genre feature averages (first 5 genres):")
print(genre_features.head().round(3))

Genre feature averages (first 5 genres):
            danceability       energy       valence       popularity      
                    mean count   mean count    mean count       mean count
track_genre                                                               
acoustic           0.549   454  0.425   454   0.424   454     42.815   454
afrobeat           0.673   433  0.700   433   0.707   433     24.372   433
alt-rock           0.537   450  0.760   450   0.533   450     31.231   450
alternative        0.550   428  0.721   428   0.488   428     24.568   428
ambient            0.370   431  0.242   431   0.169   431     44.478   431


In [21]:
# Grouped stats by explicit flag
explicit_stats = df.groupby('explicit')[['popularity', 'danceability', 
                                          'energy', 'valence']].agg(['mean', 'count'])
print("Stats by explicit flag (0=clean, 1=explicit):")
print(explicit_stats.round(3))

Stats by explicit flag (0=clean, 1=explicit):
         popularity        danceability        energy        valence       
               mean  count         mean  count   mean  count    mean  count
explicit                                                                   
0            32.948  45778        0.559  45778  0.634  45778   0.474  45778
1            36.613   4221        0.636   4221  0.720   4221   0.470   4221


In [22]:
# Grouped stats by tempo category
tempo_stats = df.groupby('tempo_category')[['popularity', 'danceability', 
                                             'energy']].agg(['mean', 'count'])
print("Stats by tempo category:")
print(tempo_stats.round(3))

Stats by tempo category:
               popularity        danceability        energy       
                     mean  count         mean  count   mean  count
tempo_category                                                    
fast               33.743  18318        0.529  18318  0.692  18318
medium             33.111  24481        0.619  24481  0.651  24481
slow               32.521   7200        0.478   7200  0.482   7200


In [8]:
# groupby with multiple aggregations — required by rubric
grouped = df.groupby(['track_genre', 'tempo_category'])['popularity'].agg(['mean', 
                                                                             'count',
                                                                             'std'])
grouped = grouped.sort_values('mean', ascending=False)
print("Top 10 genre + tempo combinations by popularity:")
print(grouped.head(10).round(3))

NameError: name 'df' is not defined

In [9]:
# Pivot table — average popularity by genre and tempo category
pivot = df.groupby(['track_genre', 'tempo_category'])['popularity'].mean()
pivot = pivot.unstack(fill_value=0).round(2)

print("Pivot table — average popularity by genre and tempo:")
print(pivot.head(10))

NameError: name 'df' is not defined

In [10]:
# Correlation matrix
corr_cols = ['popularity', 'danceability', 'energy', 'loudness', 
             'tempo', 'valence', 'acousticness', 
             'instrumentalness', 'speechiness']

corr_matrix = df[corr_cols].corr()
print("Correlation matrix:")
print(corr_matrix.round(3))

NameError: name 'df' is not defined

In [11]:
# Correlation heatmap
fig, ax = plt.subplots(figsize=(10, 8))

sns.heatmap(corr_matrix, 
            annot=True, 
            fmt='.2f',
            cmap='coolwarm',
            center=0,
            ax=ax)

ax.set_title('Correlation Matrix of Audio Features', fontsize=14)
fig.tight_layout()
fig.savefig("../figures/correlation_heatmap.png", dpi=300)
plt.show()
print("Heatmap saved!")

NameError: name 'plt' is not defined

In [12]:
# Research Q1 — What features correlate most with popularity?
popularity_corr = corr_matrix['popularity'].sort_values(ascending=False)
print("Features most correlated with popularity:")
print(popularity_corr.round(3))

NameError: name 'corr_matrix' is not defined

In [13]:
# Research Q2 — Which genre is most energetic vs most relaxed?
genre_energy = df.groupby('track_genre')['energy'].mean().sort_values(ascending=False)

print("Top 5 most energetic genres:")
print(genre_energy.head(5).round(3))
print()
print("Top 5 most relaxed genres:")
print(genre_energy.tail(5).round(3))

NameError: name 'df' is not defined

In [14]:
# Research Q3 — Do explicit songs score higher in popularity?
explicit_pop = df.groupby('explicit')['popularity'].mean()
print("Average popularity by explicit flag:")
print(explicit_pop.round(3))
print()
diff = explicit_pop[1] - explicit_pop[0]
print(f"Explicit songs score {diff:.2f} points higher on average")

NameError: name 'df' is not defined